src/libmain/progress-bar.cc

```diff
@@ -71,53 +144,220 @@ class ProgressBar : public Logger
 
-    bool printBuildLogs;
-    bool isTTY;
+    std::optional<struct termios> savedTermAttrs;
+
+    Pipe inputPipe;
+
+    const std::chrono::time_point<std::chrono::steady_clock> startTime = std::chrono::steady_clock::now();
 
 public:
 
-    ProgressBar(bool printBuildLogs, bool isTTY)
-        : printBuildLogs(printBuildLogs)
-        , isTTY(isTTY)
+    ProgressBar(bool isTTY)
+        : isTTY(isTTY)
+        , state_({ .active = isTTY })
     {
         state_.lock()->active = isTTY;
+
         updateThread = std::thread([&]() {
             auto state(state_.lock());
             while (state->active) {
+                #if 0
                 if (!state->haveUpdate)
                     state.wait(updateCV);
+                #endif
+                updateStatusLine(*state);
                 draw(*state);
                 state.wait_for(quitCV, std::chrono::milliseconds(50));
             }
         });
+
+        if (isTTY) {
+
+            struct termios term;
+            if (tcgetattr(STDIN_FILENO, &term))
+                throw SysError("getting terminal attributes");
+
+            savedTermAttrs = term;
+
+            cfmakeraw(&term);
+
+            if (tcsetattr(STDIN_FILENO, TCSANOW, &term))
+                throw SysError("putting terminal into raw mode");
+
+            inputPipe.create();
+
+            inputThread = std::thread([this]() {
+                // FIXME: exceptions
+
+                struct pollfd fds[2];
+                fds[0] = { .fd = STDIN_FILENO, .events = POLLIN, .revents = 0 };
+                fds[1] = { .fd = inputPipe.readSide.get(), .events = POLLIN, .revents = 0 };
+
+                while (true) {
+                    if (poll(fds, 2, -1) != 1) {
+                        if (errno == EINTR) continue;
+                        assert(false);
+                    }
+
+                    if (fds[1].revents & POLLIN) break;
+
+                    assert(fds[0].revents & POLLIN);
+
+                    char c;
+                    auto n = read(STDIN_FILENO, &c, 1);
+                    if (n == 0) break;
+                    if (n == -1) {
+                        if (errno == EINTR) continue;
+                        break;
+                    }
+                    c = std::tolower(c);
+
+                    if (c == 3 || c == 4 || c == 'q') {
+                        auto state(state_.lock());
+                        state->statusLines.insert_or_assign({idQuit, 0}, ANSI_RED "Exiting...");
+                        draw(*state);
+                        triggerInterrupt();
+                    }
+
+                    {
+                        auto state(state_.lock());
+                        if (state->prompt) {
+                            state->prompt->set_value(c != '\n' ? c : std::optional<char>());
+                            state->prompt.reset();
+                            continue;
+                        }
+                    }
+
+                    if (c == 'l') {
```

Comment:
These `c == 'x'` switches should probably have an explicit `else if` for all the other cases. That makes it explicit that only one of them will be active at a time. Making reading this code easier and more explicit. Another option would be a switch on `c`.